### PR TITLE
The true ruination of space: Death to the donut, let thy cube be done [wip]

### PR DESCRIPTION
**Pull request by Incoming5643**
All existing structures in space outside of mining/lavaland, the station, and centcom have been converted to ruins. Every z level that formally contained structures now spawns ruins instead. There's no assurances that anything in space will or... willn't exist in any round.

Sole exception right now is the deep space dock for the white ship, because that kind of NEEDS to exist.

z10 has been added to maps with lavaland to make the number of crosslinked z levels up to a solid 9.

z10 is the "true deep space" level where nothing can ever spawn, this is mostly because it doesn't exist in some maps.

Todo:
- [ ] Catch all those horrible things that rely on the max number of z levels to update them
- [ ] Update map transitions so it always forms a 3 by 3 grid, instead of a 9 by 1 line
- [ ] Figure out which issues this fixes for MAD TOKENS

### CODE REVIEW

**Comment by optimumtact** [comment]
+2,762 −206,606 :scream:

**Comment by Incoming5643** [comment]
blanking out those z-levels built up a powerful thirst

**Comment by Bawhoppen** [comment]
will the whiteship docking port on the derelict work? also wouldn't it be better if we just made the whiteship not spawn on z1?

**Comment by Incoming5643** [comment]
the benefit of having the white ship start docked on the station is that every station can have a different white ship if they want it. Though I could make both white ships ruins then we'd have to be sure they both didn't spawn at the same time.

I have no idea if the derelict white ship port works, it depends if the white ship looks for ports before or after ruin generation, but my hope is it looks after, in which case it works just fine.

**Comment by Core0verload** [comment]
Do you even know how fucking long it takes for our map loader to load a single map? A single _small_ map? Now scale it up for every single space derelict we have, including КС13 and old AI sat. Oh fuck.

You are _ruining_ any hope for us to have a short server startup time. And it makes local testing even more of pain in ass than it already is.

**Comment by Incoming5643** [comment]
Lavaland works off a config budget, if you set it to 0 you can have a server that never spawns lavaland ruins.

I can do the same sort of thing here for space so it'd be easy to set the station not to bother with ruin loading at all.

As for testing specific ruins, you can load those in from the game.

**Comment by optimumtact** [comment]
We can't keep optimising for every man and his toaster server.

**Comment by Incoming5643** [comment]
It literally takes me over half  an hour to go from pressing the compile button to being in a test server of my own creation.

I know this pain all too well.

**Comment by optimumtact** [comment]
HALF AN HOUR?

WHAT KIND OF SHITBOX ARE YOU USING?

**Comment by Incoming5643** [comment]
THE LATEST IN 2008 LAPTOP TECHNOLOGY
https://support.lenovo.com/us/en/documents/pd008989

but we're getting off track. I can't reasonably tell you how much time this will add to start up in a real server, but I suspect the answer is: Not much.

**Comment by Core0verload** [comment]
>We can't keep optimising for every man and his toaster server.

This is how you get a painfully long startup time on /tg/ _main server_. By merging hundreds of small startup de-optimizations. "Nah, tg server can take it." Atmos refactor is +50 seconds flat, and ruin load delay, when measured, would be somewhat close.

**Comment by Incoming5643** [comment]
When this is complete we can test merge it and get some real numbers™ to see how real the argument is.

**Comment by Core0verload** [comment]
You can measure load times locally, and compare it to light setup to get a feel of how much it would impact the server. No need for test merge.

**Comment by Core0verload** [comment]
ah, and before I forget, there is one more problem with ruins

**path updates**

nobody ever updates paths on ruins, and we have no travis set up to do map checks on them

this problem already exists, but turning more things into ruins makes it worse

**Comment by Supermichael777** [comment]
potato+screen=server

compiling takes a good 10 min and actually getting into the lobby takes a hard 4 min.

used to only take 30+\- seconds to load but then lava happened. also enabling 7 overlays for every mob/living was probably a bad idea

**Comment by KorPhaeron** [comment]
Ruins are definitely a maintainability issue. Half of them already dont work because of the turf repaths

**Comment by Incoming5643** [comment]
Neither does the default mining station. Watching repaths in general is just an issue. People just gotta be more honest when they repath and/or someone's gotta make a tool that can automatically fix repaths when they change.

**Comment by Core0verload** [comment]
Just add a new test to Travis - make it try to compile with all the template maps checked. It would make getting broken repaths merged harder.

**Comment by Iamgoofball** [comment]
@Core0verload piss easy

also, the main problems with server startup times that Lummox brought up is list init() stuff

**Comment by Core0verload** [comment]
Maybe, but +50s subsystems time from atmos refactor is still noticeable.

**Comment by Iamgoofball** [comment]
Wait for stonedmc

**Comment by Core0verload** [comment]
New MC incudes a fix for lights init?

**Comment by Iamgoofball** [comment]
light and atmos init are on his top priority list of things to speed up

**Comment by Razharas** [comment]
How to make it lump together instead of completely randomized setup:

Get a sqrt of a number of linked zlevels
Round it to upper
Make a square grid with the side = previously calculated number
Fill the grid with the point datums randomly
Run the linking algorythm that will do all the work itself
It will have holes in it if you have a non-rootable amount of zlevels, if its 9, 16, 25 and so on it will be uniform with no unusual transitions

**Comment by Razharas** [comment]
Some procs in points may need adjustments though

**Comment by xxalpha** [comment]
So now we have even longer loading times and there won't always be the clown ship or the derelict?

**Comment by Incoming5643** [comment]
I can force some ruins to always load, but the idea is true randomness if I can help it.

**Comment by Bawhoppen** [comment]
why don't we just implement fixed grid nowadays then if everything is randomized

**Comment by Incoming5643** [comment]
it's still nominally important to randomize the z levels in relation to one another because otherwise you could tell your bearing from the station by which z level you were on. Like if it was always:

`3 4 6`
`7 1 8`
`9 10 11`

and you saw you were on 9 you'd know you could get home if you threw something that sent you north east. (also just realized I need z11 too)

People could argue that's a good thing though.

**Comment by xxalpha** [comment]
Does this even fix the infinite loop?

**Comment by Incoming5643** [comment]
if it's rendered in a 3 by 3 grid yes.

Space torus seems to almost exclusively render in a X by 1 line, which leaves one direction infinite

**Comment by xxalpha** [comment]
I don't see the point in changing all this if the infinite loop still happens.

**Comment by Bawhoppen** [comment]
No but like the crosslinks randomize

ss 2016-06-03 at 03 20 26

**Comment by Razharas** [comment]
In current way transitions are done its impossiblke to have infinite loop, its possible to have a long one, but never infinite

**Comment by Bawhoppen** [comment]
thats not true

**Comment by Razharas** [comment]
It is true

**Comment by Bawhoppen** [comment]
it isn't I have proof

**Comment by Razharas** [comment]
Post it

**Comment by Bawhoppen** [comment]
go onto server, find the just right spot to jump off, repeatedly pass the same landmark for eternity

**Comment by Bawhoppen** [comment]
(right spot is random and sometimes doesnt exist but sometimes does)

**Comment by Razharas** [comment]
By right spot do you mean somewhere in space?
Because current settings are supposed to return you to the location you started from eventually
If you started in the empty space and went to the side that is looped back on itself you will be in an endless loop, but that would be self-imposed one

**Comment by xxalpha** [comment]
How is it self-imposed, you can't expect players to avoid infinite loops, they shouldn't happen period.

**Comment by Razharas** [comment]
It is self imposed because you are returning to the place you started your movement from, if you start from empty area and it just so happens that there is nothing in the direction you are going you will return to that empry area
Its impossible to make endless loops never happen, in the worst case singulo can eat all of the staiton and asteroids and space content and every movement will become an endless loop

**Comment by Bawhoppen** [comment]
its still an endless loop

**Comment by Razharas** [comment]
Endless loops of such type are impossible to avoid, this pr will not fix it either

**Comment by Bawhoppen** [comment]
this pr wasn't trying to fix eiher though?

**Comment by Razharas** [comment]
Incoming claimed that 3x3 grid will fix the infinite loop
It will not fix the possibility of infinite loop, it will just make space map very simple, thus making players impose the infinite space drift upon themselves less likely

**Comment by Razharas** [comment]
Heres the explanation, red dot is a player, small arrow near him is the direction of drift, 3 strangely shaped square lumps are possible space maps
As you can see no matter how the maps are composed this dude will drift forever because there is nothing on the same x-titude as him on any map
But as you can also see only way to be on this x-titude and have such a drift direction is if you throw something while drifting or someone carry you on that x-titude and throws you, you cant enter this infinite loop from station airlock by accident, thats why such an infinite loop is self(or enemy) imposed
Its impossible to get rid of such loops because as i said, in worst case singulo can eat everything and you will never arrive to anything ever
image

**Comment by Bawhoppen** [comment]
not if we use the grids like I'm suggesting, as long as we make the offset be a specific size they could theoretically hit every single coordinate in space

**Comment by Incoming5643** [comment]
The only thing I'm trying to get rid of on the large scale structure of space are map layouts that are 1 by X chains of z levels that have self loops.

File: code/datums/ruins/space.dm [hunk 1]
```diff
@@ -155,3 +155,66 @@
...
+	name = "Syndicate Listening Station"
+	description = "Listening stations form the backbone of the syndicate's information gathering operations. \
+		Assignment to these stations is dreaded by most agents, as it entails long and lonely shifts \
+		listening to nearby stations chatter incessently about the most meaningless things."
+
+/datum/map_template/ruin/space/oldAIsat
+	id = "oldAIsat"
+	suffix = "oldAIsat.dmm"
+	name = "Abandoned AI Satellite"
+	description = "Early experiments in AI housing focused on large remote structures that could not be easily penitrated. \
```
Comment (coiax) [inline code/datums/ruins/space.dm]: penetrated

File: code/datums/ruins/space.dm [hunk 2]
```diff
@@ -155,3 +155,66 @@
...
+		Over time the benifits of having immediate physicall access to the AI have been proven in the blood of dozens of \
```
Comment (coiax) [inline code/datums/ruins/space.dm]: benefits, physical

File: code/datums/ruins/space.dm [hunk 3]
```diff
@@ -155,3 +155,66 @@
...
+	id = "djstation"
+	suffix = "djstation.dmm"
+	name = "DJ Station"
+	description = "Until very recently this pirate radio station was used \
+		to harangue local space stations over a variety of perceived \"ethics violations\". \
+		It seems like someone finally got sick of it, but the equipment still works."
+
+/datum/map_template/ruin/space/thederelict
+	id = "thederelict"
+	suffix = "thederelict.dmm"
```
Comment (coiax) [inline code/datums/ruins/space.dm]: It's named TheDerelict.dmm in the tree, but lowercase in this entry

File: code/datums/ruins/space.dm [hunk 4]
```diff
@@ -155,3 +155,66 @@
...
+	was outfited with a multitude of food generating features, then summarily ran into an asteroid shortly \
+		after takeoff."
+
+/datum/map_template/ruin/space/listeningstation
+	id = "listeningstation"
+	suffix = "listeningstation.dmm"
+	name = "Syndicate Listening Station"
+	description = "Listening stations form the backbone of the syndicate's information gathering operations. \
+		Assignment to these stations is dreaded by most agents, as it entails long and lonely shifts \
+		listening to nearby stations chatter incessently about the most meaningless things."
```
Comment (coiax) [inline code/datums/ruins/space.dm]: incessantly

File: code/datums/ruins/space.dm [hunk 5]
```diff
@@ -155,3 +155,66 @@
...
+	name = "Crashed Ship"
+	description = "Among civilian vessels the most common cause of tragedy is lack of food. This ship \
+	was outfited with a multitude of food generating features, then summarily ran into an asteroid shortly \
+		after takeoff."
+
+/datum/map_template/ruin/space/listeningstation
+	id = "listeningstation"
+	suffix = "listeningstation.dmm"
+	name = "Syndicate Listening Station"
+	description = "Listening stations form the backbone of the syndicate's information gathering operations. \
```
Comment (coiax) [inline code/datums/ruins/space.dm]: Syndicate

File: code/datums/ruins/space.dm [hunk 6]
```diff
@@ -155,3 +155,66 @@
...
+	description = "In space construction the teleporter is often the first system brought online. \
+		This lonely half built teleporter is a sign of a proposed structure that for one reason or another \
+		just never got built."
+
+/datum/map_template/ruin/space/crashedclownship
+	id = "crashedclownship"
+	suffix = "crashedclownship.dmm"
+	name = "Crashed Clown Ship"
+	description = "For centuries the promise of a new clown homeworld has been the siren call for countless \
+		clown vessels. Alas the clown's lust for shinanagans means that successful voyages are almost unheard of, \
```
Comment (coiax) [inline code/datums/ruins/space.dm]: shenanigans

**Comment by coiax** [comment]
Let me work on the white ship. I have an idea.

**Comment by Razharas** [comment]
Making a skewed transitions between maps would if anything make space navigation even harder to comprehend

File: code/datums/ruins/space.dm (hunk 1 above)
Comment (PKPenguin321) [inline code/datums/ruins/space.dm]: ;)

File: code/datums/ruins/space.dm (hunk 3 above)
Comment (Incoming5643) [inline code/datums/ruins/space.dm]: Every time I rename it in dreammaker it reverts to capitals it's the weirdest thing.

**Comment by Incoming5643** [comment]
carry me to the finish line @coiax 

#18484

**Pull request closed without merging**